### PR TITLE
Add support for multipart parameters

### PR DESCRIPTION
**Pull request by mikegoatly**
Referencing issue #19 

These are some quick thoughts on how multipart content could be supported in webanchor.

Basic usage:
```csharp
        [BaseLocation("api")]
        public interface IApi
        {
            [Post("")]
            Task<HttpResponseMessage> PostMultipart([Content][Multipart]MultipartContentData data);
        }

        [Fact]
        public void TestPostWithStringContent()
        {
            TestTheRequest<IApi>(
               api => api.PostMultipart(new MultipartContentData(new StringContentPart("Content", "Field1", "test.txt") { ContentType = "text/plain" })),
               assertMe =>
               {
                   Assert.Equal(HttpMethod.Post, assertMe.Method);
                   Assert.Equal("api", assertMe.RequestUri.ToString());
                   var body = assertMe.Content.ReadAsStringAsync().Result;
                   Assert.True(Regex.IsMatch(body, @"--(?<id>[0-9a-f\-]*)
Content-Type: text/plain
Content-Disposition: form-data; name=Field1; filename=test\.txt; filename\*=utf-8''test\.txt

Content
--\k<id>--"));
               });
        }
```

Where you can add as many parts as you want to a MultipartContentData parameter, with support for `StringPart`s, `ByteArrayPart`s and `StreamPart`s.

Adding a new [Multipart] attribute to the parameter plays nicely with the plug-in system you've built for adapting the parameter transformer (it could just as well be on the method though), but it does feel a bit redundant because of the way I'm expecting a `MultipartContentData` as the parameter type.

I also wasn't sure where you'd want the `...ContentPart` types living in the project hierarchy.

Interested in your thoughts.

### CODE REVIEW

File: WebAnchor/ContentPart.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using System.Net.Http;
+using System.Net.Http.Headers;
+
+namespace WebAnchor
+{
+    public abstract class ContentPartBase
```
Comment (mattiasnordqvist) [inline WebAnchor/ContentPart.cs]: Change file name of file to match class name.

File: WebAnchor/MultipartContent.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Collections.Generic;
+
+namespace WebAnchor
+{
+    public class MultipartContentData
```
Comment (mattiasnordqvist) [inline WebAnchor/MultipartContent.cs]: Change file name to match class name

**Comment by mattiasnordqvist** [comment]
Experienced web anchor users @spinit-moom and @carl-berg, do you have any input

**Comment by mikegoatly** [comment]
Great feedback, thanks. 

I really like the idea of a paramarray instead of the MultipartContentCata type - it removes a bit of unnecessary plumbing. It only took a couple of minutes to do, but already this feels a bit cleaner:

``` csharp
[BaseLocation("api")]
public interface IApi
{
    [Post("")]
    Task<HttpResponseMessage> PostSingleMultipart([Content][Multipart]ContentPartBase data);
    
    [Post("")]
    Task<HttpResponseMessage> PostParamArrayMultipart([Content][Multipart]params ContentPartBase[] data);
    
    [Post("")]
    Task<HttpResponseMessage> PostListMultipart([Content][Multipart]List<ContentPartBase> data);
}
```

And usage is obviously more concise too:
```csharp
 api.PostParamArrayMultipart(
    new StringContentPart("Content", "Field1", "test.txt") { ContentType = "text/plain" },
    new ByteArrayContentPart(Encoding.UTF8.GetBytes("From bytes"), "Field1", "test.txt") { ContentType = "text/plain" })
```

Any parameter that is castable to IEnumerable<ContentPartBase> is acceptable (hence the paramarray and List<> examples), but we can also allow for a single content part purely for convenience when only one part needs to be uploaded - not entirely necessary, but reduces a bit of syntax.

**Comment by mikegoatly** [comment]
Apologies if I've jumped the gun on making the changes :smile:

**Comment by mattiasnordqvist** [comment]
I think this looks great! :+1:

**Comment by mattiasnordqvist** [comment]
On it's way https://www.nuget.org/packages/WebAnchor/6.2.0-pre1

**Pull request merged**